evaluation_service/src/main.rs

```diff
@@ -1,80 +1,590 @@
+                })
+            })
+    }
+
+    fn get_evaluation_file_resource_name(&self, eval_type: evaluation_file::Type) -> &str {
+        match eval_type {
+            evaluation_file::Type::Checker => "checker",
+            evaluation_file::Type::Interactor => "interactor",
+        }
+    }
+}
 
 #[tonic::async_trait]
 impl Evaluation for EvaluationService {
     async fn get_user_scoring(
         &self,
         _request: Request<GetUserScoringRequest>,
     ) -> Result<Response<GetUserScoringResponse>, Status> {
-        todo!()
+        self.storage
+            .search_item(None, USER_SCORING_FILE_NAME, Some(SERIALIZED_EXTENSION))
+            .map_err(not_found_error)
+            .and_then(|op| op.ok_or_else(|| not_found_error("User scoring method not found")))
+            .and_then(|path| {
+                self.storage
+                    .read_file_object(&path)
+                    .map_err(|err| internal_error(err.as_ref()))
+            })
+            .map(|user| Response::new(GetUserScoringResponse { method: user }))
     }
 
     async fn get_problem(
         &self,
-        _request: Request<GetProblemRequest>,
+        request: Request<GetProblemRequest>,
     ) -> Result<Response<GetProblemResponse>, Status> {
-        todo!()
-    }
-
-    async fn set_contest(
-        &self,
-        _request: Request<SetContestRequest>,
-    ) -> Result<Response<SetContestResponse>, Status> {
-        todo!()
+        let request = request.into_inner();
+        self.get_problem_folder_from_id(request.problem_id)
+            .map_err(internal_error)
+            .and_then(|path| {
+                self.storage
+                    .search_item(
+                        Some(&path),
+                        PROBLEM_METADATA_FILE_NAME,
+                        Some(SERIALIZED_EXTENSION),
+                    )
+                    .map_err(not_found_error)
+            })
+            .and_then(|op| {
+                op.ok_or_else(|| {
+                    not_found_error(format!(
+                        "Problem metadata not found [id: {}]",
+                        request.problem_id
+                    ))
+                })
+            })
+            .and_then(|path| {
+                self.storage
+                    .read_file_object(&path)
+                    .map_err(|err| internal_error(err.as_ref()))
+            })
+            .map(|prob| Response::new(GetProblemResponse { info: prob }))
     }
 
     async fn get_contest(
         &self,
         _request: Request<GetContestRequest>,
     ) -> Result<Response<GetContestResponse>, Status> {
-        todo!()
+        let user_scoring_method = self
+            .get_user_scoring(Request::new(GetUserScoringRequest {}))
+            .await?
+            .into_inner()
+            .method;
+        let mut problems: Vec<Problem> = vec![];
+        for entry in self.storage.iterate_folder(PROBLEMS_FOLDER_NAME, None)? {
+            let problem_path = self
```

Comment:
Cannot do it since my closure yields a result.
There are ways to do it in a more functional way, but it involves the use of `process_results`in the `itertools`create.

Reply:
Yeah I was focusing on the vector building part because you don’t need the vector to be mutable